File: python/src/starline/smoothness.py
"""Smoothness engine: jerk + curvature alignment."""

from __future__ import annotations

import math

from .config import EngineConfig
from .player import PlayerState
from .spline import SplineTrack


def steering_response(p: PlayerState, cfg: EngineConfig, kappa: float) -> float:
    """Steering maps to a target curvature response (geometry feel)."""
    target = p.steer * cfg.steering_gain * max(abs(kappa), 1e-6)
    p.steering_response = target
    return target


def curve_error(p: PlayerState, kappa: float, steer_resp: float) -> float:
    p.curve_error = abs(kappa - steer_resp)
    return p.curve_error


def anticipatory_kappa(
    track: SplineTrack,
    s: float,
    v_eff: float,
    cfg: EngineConfig,
) -> tuple[float, float]:
    """κ_now and κ_ahead = κ(s + Δs) for predictive smoothness."""
    ds = cfg.anticipation_ds + cfg.anticipation_speed_scale * max(v_eff, 0.0)
    k0 = track.curvature(s)
    k1 = track.curvature(s + ds)
    return k0, k1


def smoothness_score(
    p: PlayerState,
    cfg: EngineConfig,
    kappa: float,
    kappa_ahead: float | None = None,
) -> float:
    """S = exp(- (w_j*|jerk| + w_c*curve_error)).

    If kappa_ahead provided, blend reactive + anticipatory alignment.
    """
    w_a = cfg.anticipation_blend if kappa_ahead is not None else 0.0
    k_eff = (1.0 - w_a) * kappa + w_a * float(kappa_ahead) if kappa_ahead is not None else kappa

    steer_resp = steering_response(p, cfg, k_eff)
    cerr = curve_error(p, kappa, steer_resp)
    if kappa_ahead is not None and w_a > 0.0:
        steer_a = steering_response(p, cfg, kappa_ahead)
        cerr = (1.0 - w_a) * cerr + w_a * curve_error(p, kappa_ahead, steer_a)

    penalty = cfg.jerk_weight * abs(p.jerk) + cfg.curve_weight * cerr
    p.smoothness = math.exp(-penalty)
    return p.smoothness


def instability(p: PlayerState, cfg: EngineConfig) -> float:
    """Sudden steering / braking spikes."""
    dsteer = abs(p.steer - p.steer_prev)
    p.steer_prev = p.steer
    inst = (
        cfg.steer_instability_weight * dsteer
        + cfg.brake_instability_weight * p.brake
    )
    p.instability = inst
    return inst

File: roblox/Starline/Smoothness.lua
--!strict
local Config = require(script.Parent.Config)

local Smoothness = {}

local function C(cfg: any?): any
	return cfg or Config
end

function Smoothness.steeringResponse(state: any, kappa: number, cfg: any?): number
	local c = C(cfg)
	local resp = state.steer * c.SteeringGain * math.max(math.abs(kappa), 1e-6)
	state.steeringResponse = resp
	return resp
end

function Smoothness.curveError(state: any, kappa: number, steerResp: number): number
	local err = math.abs(kappa - steerResp)
	state.curveError = err
	return err
end

function Smoothness.anticipatoryKappa(track: any, s: number, vEff: number, cfg: any?): (number, number)
	local c = C(cfg)
	local ds = c.AnticipationDs + c.AnticipationSpeedScale * math.max(vEff, 0)
	local k0 = track:curvature(s)
	local k1 = track:curvature(s + ds)
	return k0, k1
end

function Smoothness.score(state: any, kappa: number, kappaAhead: number?, cfg: any?): number
	local c = C(cfg)
	local wA = if kappaAhead ~= nil then c.AnticipationBlend else 0
	local steerResp = Smoothness.steeringResponse(state, kappa, c)
	local cerr = Smoothness.curveError(state, kappa, steerResp)
	if kappaAhead ~= nil and wA > 0 then
		local steerA = Smoothness.steeringResponse(state, kappaAhead, c)
		local cerrA = Smoothness.curveError(state, kappaAhead, steerA)
		cerr = (1 - wA) * cerr + wA * cerrA
	end
	local penalty = c.JerkWeight * math.abs(state.jerk) + c.CurveWeight * cerr
	state.smoothness = math.exp(-penalty)
	return state.smoothness
end

function Smoothness.instability(state: any, cfg: any?): number
	local c = C(cfg)
	local dsteer = math.abs(state.steer - state.steerPrev)
	state.steerPrev = state.steer
	local inst = c.SteerInstabilityWeight * dsteer + c.BrakeInstabilityWeight * state.brake
	state.instability = inst
	return inst
end

return Smoothness
